File: DSA/DSA_450/Cpp/07_binary_search_trees/015_Median_Of_BST.cpp
/*
Problem: Find Median of BST
URL: https://www.geeksforgeeks.org/find-median-bst-time-o1-space/

Problem Statement:
Find median of BST in O(n) time and O(1) space.

Sample Input/Output:
Input: root = [6,3,8,1,5,7,9]
Output: 6
Explanation: Inorder: [1,3,5,6,7,8,9], median is 6.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int data;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : data(x), left(NULL), right(NULL) {}
};

TreeNode* Insert_BST(TreeNode* root, int key) {
    if (root == NULL) return new TreeNode(key);
    if (key < root->data) root->left = Insert_BST(root->left, key);
    else root->right = Insert_BST(root->right, key);
    return root;
}

void Print_Inorder(TreeNode* root) {
    if (root == NULL) return;
    Print_Inorder(root->left);
    cout << root->data << " ";
    Print_Inorder(root->right);
}

class Solution {
public:
    int Count_Nodes(TreeNode* root) {
        int count = 0;
        TreeNode* curr = root;
        while (curr != NULL) {
            if (curr->left == NULL) {
                count++;
                curr = curr->right;
            } else {
                TreeNode* prev = curr->left;
                while (prev->right != NULL && prev->right != curr) {
                    prev = prev->right;
                }
                if (prev->right == NULL) {
                    prev->right = curr;
                    curr = curr->left;
                } else {
                    prev->right = NULL;
                    count++;
                    curr = curr->right;
                }
            }
        }
        return count;
    }

    int Find_Kth_Node_Morris(TreeNode* root, int k) {
        TreeNode* curr = root;
        int count = 0;
        while (curr != NULL) {
            if (curr->left == NULL) {
                count++;
                if (count == k) return curr->data;
                curr = curr->right;
            } else {
                TreeNode* prev = curr->left;
                while (prev->right != NULL && prev->right != curr) {
                    prev = prev->right;
                }
                if (prev->right == NULL) {
                    prev->right = curr;
                    curr = curr->left;
                } else {
                    prev->right = NULL;
                    count++;
                    if (count == k) return curr->data;
                    curr = curr->right;
                }
            }
        }
        return -1;
    }

    double Median_Morris(TreeNode* root) {
        /*
        Morris traversal with node counting approach
        Time Complexity: O(n)
        Space Complexity: O(1)
        */
        int n = Count_Nodes(root);
        if (n % 2 == 1) {
            return Find_Kth_Node_Morris(root, (n + 1) / 2);
        } else {
            int first = Find_Kth_Node_Morris(root, n / 2);
            int second = Find_Kth_Node_Morris(root, n / 2 + 1);
            return (first + second) / 2.0;
        }
    }

    void Inorder_To_Array(TreeNode* root, vector<int>& arr) {
        if (root == NULL) return;
        Inorder_To_Array(root->left, arr);
        arr.push_back(root->data);
        Inorder_To_Array(root->right, arr);
    }

    double Median_Array(TreeNode* root) {
        /*
        Inorder to array approach
        Time Complexity: O(n)
        Space Complexity: O(n)
        */
        vector<int> arr;
        Inorder_To_Array(root, arr);
        int n = arr.size();
        if (n % 2 == 1) {
            return arr[n / 2];
        } else {
            return (arr[n / 2 - 1] + arr[n / 2]) / 2.0;
        }
    }
};

void Test_Median_Of_BST() {
    Solution solution;
    TreeNode* root = NULL;
    root = Insert_BST(root, 6);
    root = Insert_BST(root, 3);
    root = Insert_BST(root, 8);
    root = Insert_BST(root, 1);
    root = Insert_BST(root, 5);
    root = Insert_BST(root, 7);
    root = Insert_BST(root, 9);
    cout << "Median (Morris): " << solution.Median_Morris(root) << endl;
    cout << "Median (Array): " << solution.Median_Array(root) << endl;
}

int main() {
    Test_Median_Of_BST();
    return 0;
}

File: DSA/DSA_450/Python/07_binary_search_trees/015_Median_Of_BST.py
"""
Problem: Find Median of BST
URL: https://www.geeksforgeeks.org/find-median-bst-time-o1-space/

Problem Statement:
Find median of BST in O(n) time and O(1) space.

Sample Input/Output:
Input: root = [6,3,8,1,5,7,9]
Output: 6
Explanation: Inorder: [1,3,5,6,7,8,9], median is 6.
"""


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    @staticmethod
    def Build_BST(keys):
        root = None
        for key in keys:
            root = TreeNode.Insert_BST(root, key)
        return root

    @staticmethod
    def Insert_BST(root, key):
        if root is None:
            return TreeNode(key)
        if key < root.val:
            root.left = TreeNode.Insert_BST(root.left, key)
        else:
            root.right = TreeNode.Insert_BST(root.right, key)
        return root

    @staticmethod
    def Print_Inorder(root):
        if root is None:
            return
        TreeNode.Print_Inorder(root.left)
        print(root.val, end=" ")
        TreeNode.Print_Inorder(root.right)


class Solution:
    def Count_Nodes(self, root):
        count = 0
        curr = root
        while curr is not None:
            if curr.left is None:
                count += 1
                curr = curr.right
            else:
                prev = curr.left
                while prev.right is not None and prev.right != curr:
                    prev = prev.right
                if prev.right is None:
                    prev.right = curr
                    curr = curr.left
                else:
                    prev.right = None
                    count += 1
                    curr = curr.right
        return count

    def Find_Kth_Node_Morris(self, root, k):
        curr = root
        count = 0
        while curr is not None:
            if curr.left is None:
                count += 1
                if count == k:
                    return curr.val
                curr = curr.right
            else:
                prev = curr.left
                while prev.right is not None and prev.right != curr:
                    prev = prev.right
                if prev.right is None:
                    prev.right = curr
                    curr = curr.left
                else:
                    prev.right = None
                    count += 1
                    if count == k:
                        return curr.val
                    curr = curr.right
        return -1

    def Median_Morris(self, root):
        """
        Morris traversal with node counting approach
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        n = self.Count_Nodes(root)
        if n % 2 == 1:
            return self.Find_Kth_Node_Morris(root, (n + 1) // 2)
        else:
            first = self.Find_Kth_Node_Morris(root, n // 2)
            second = self.Find_Kth_Node_Morris(root, n // 2 + 1)
            return (first + second) / 2.0

    def Inorder_To_Array(self, root, arr):
        if root is None:
            return
        self.Inorder_To_Array(root.left, arr)
        arr.append(root.val)
        self.Inorder_To_Array(root.right, arr)

    def Median_Array(self, root):
        """
        Inorder to array approach
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        arr = []
        self.Inorder_To_Array(root, arr)
        n = len(arr)
        if n % 2 == 1:
            return arr[n // 2]
        else:
            return (arr[n // 2 - 1] + arr[n // 2]) / 2.0


def Test_Median_Of_BST():
    solution = Solution()
    root = None
    root = TreeNode.Insert_BST(root, 6)
    root = TreeNode.Insert_BST(root, 3)
    root = TreeNode.Insert_BST(root, 8)
    root = TreeNode.Insert_BST(root, 1)
    root = TreeNode.Insert_BST(root, 5)
    root = TreeNode.Insert_BST(root, 7)
    root = TreeNode.Insert_BST(root, 9)
    print("Median (Morris):", solution.Median_Morris(root))
    print("Median (Array):", solution.Median_Array(root))


if __name__ == "__main__":
    Test_Median_Of_BST()
